package/health_daemon/src/base_num_sats_monitor.c

```diff
@@ -0,0 +1,213 @@
+/*
+ * Copyright (C) 2018 Swift Navigation Inc.
+ * Contact: Swift Navigation <dev@swiftnav.com>
+ *
+ * This source is subject to the license found in the file 'LICENSE' which must
+ * be be distributed together with this source. All other rights reserved.
+ *
+ * THIS CODE AND INFORMATION IS PROVIDED "AS IS" WITHOUT WARRANTY OF ANY KIND,
+ * EITHER EXPRESSED OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE IMPLIED
+ * WARRANTIES OF MERCHANTABILITY AND/OR FITNESS FOR A PARTICULAR PURPOSE.
+ */
+
+/**
+ * \file base_num_sats_monitor.c
+ * \brief Base Observations Number of Satellites Health Monitor
+ *
+ * Monitors the unique satellites that are being supplied by the base station,
+ * and alerts when that number is below a viable solution threshold.
+ * Will not alert when not connected to a base station (no base
+ * obs messages being received).
+ * \author Ben Altieri
+ * \version v2.2.0
+ * \date 2018-09-20
+ */
+
+#include <stdlib.h>
+#include <string.h>
+
+#include <libpiksi/logging.h>
+
+#include <libsbp/sbp.h>
+#include <libsbp/observation.h>
+
+#include <swiftnav/signal.h>
+#include <swiftnav/single_epoch_solver.h>
+
+#include "health_monitor.h"
+
+#include "glo_health_context.h"
+#include "base_num_sats_monitor.h"
+
+/* these are from fw private, consider moving to libpiski */
+#define MSG_FORWARD_SENDER_ID (0u)
+
+#define BASE_NUM_SATS_SAMPLE_RATE (10000u)      /* ms */
+#define BASE_NUM_SATS_ALERT_RATE_LIMIT (30000u) /* ms */
+
+/**
+ * \brief Private context for the glo obs health monitor
+ */
+static health_monitor_t *base_num_sats_monitor;
+
+/******* Local Context Helpers Begin ********/
+
+static u8 max_timeout_count_before_warn =
+  (u8)(BASE_NUM_SATS_ALERT_RATE_LIMIT / BASE_NUM_SATS_SAMPLE_RATE);
+
+/**
+ * \brief Private global data for base num sats callbacks
+ */
+static struct base_num_sats_ctx_s {
+  gnss_signal_t unique_sids[MIN_SATS_FOR_PVT];
+  u8 num_sats;
+  u8 timeout_counter;
+} base_num_sats_ctx = {0};
+
+static bool base_num_sats_below_threshold(void)
+{
+  return (base_num_sats_ctx.num_sats < MIN_SATS_FOR_PVT);
+}
+
+static void base_num_sats_reset(void)
+{
+  memset(base_num_sats_ctx.unique_sids, 0, sizeof(base_num_sats_ctx.unique_sids));
+  base_num_sats_ctx.num_sats = 0;
+}
+
+static void base_num_sats_ingest_sid(const gnss_signal_t sid)
+{
+  // Don't need to check unless we're below the threshold
+  if (!base_num_sats_below_threshold()) return;
+
+  for (int i = 0; i < base_num_sats_ctx.num_sats; i++) {
+    if (sid_compare(sid, base_num_sats_ctx.unique_sids[i]) == 0) {
+      // We've found a duplicate, so ignore
+      return;
+    }
+  }
+
+  // Not found in current set, so add and increment num_sats
+  base_num_sats_ctx.unique_sids[base_num_sats_ctx.num_sats++] = sid;
```

Comment:
It's a bit unclear are we counting number of unique satellites or unique signals. Nomenclature indicates unique satellites but functionality looks like unique signals. You might want to check `swiftnav/sid_set.h`, I think it could be used here.

Reply:
@benjaminaltieri  We should count the number of unique sats (not signals)

Reply:
Ok great, thanks for that. Honestly I was taking a crack at this without fully understanding the way to determine a 'unique satellite'. So would sampling within a particular interval and looking for the number of sats per period makes sense? If so, is ten seconds an appropriate amount of time? If not, should it be per epoch? Or should I try to keep a running track of the sids and manually remove those that aren't seen anymore?

Reply:
Am I basically duplicating this check?

https://github.com/swift-nav/starling/blob/9a6334c3bb08749fc651f80c9d90d70bfb31b6c2/src/starling/starling_obs_converter.c#L39-L49